contrib/bpftrace/taskqlatency.sh

```
@@ -0,0 +1,67 @@
+#!/bin/bash
+
+read -r -d '' PROGRAM <<'EOF'
+#include <sys/taskq.h>
+
+kprobe:trace_zfs_taskq_ent__birth
+{
+        $tqent = (struct taskq_ent *)arg0;
+
+        $tqent_id = $tqent->tqent_id;
+        $tq_name = str($tqent->tqent_taskq->tq_name);
+
+        @birth[$tq_name, $tqent_id] = nsecs;
+}
+
+kprobe:trace_zfs_taskq_ent__start
+{
+        $tqent = (struct taskq_ent *)arg0;
+
+        @tqent_id[tid] = $tqent->tqent_id;
+        @tq_name[tid] = str($tqent->tqent_taskq->tq_name);
+
+        @start[@tq_name[tid], @tqent_id[tid]] = nsecs;
+}
+
+kprobe:trace_zfs_taskq_ent__finish
+/ @birth[@tq_name[tid], @tqent_id[tid]] /
+{
+        @queue_lat_us[@tq_name[tid]] =
+                hist((nsecs - @birth[@tq_name[tid], @tqent_id[tid]]));
+        delete(@birth[@tq_name[tid], @tqent_id[tid]]);
+}
+
+kprobe:trace_zfs_taskq_ent__finish
+/ @start[@tq_name[tid], @tqent_id[tid]] /
+{
+        $tqent = (struct taskq_ent *)arg0;
+
+        @exec_lat_us[@tq_name[tid], ksym($tqent->tqent_func)] =
+                hist((nsecs - @start[@tq_name[tid], @tqent_id[tid]]));
+        delete(@start[@tq_name[tid], @tqent_id[tid]]);
+}
+
+kprobe:trace_zfs_taskq_ent__finish
+{
+        delete(@tq_name[tid]);
+        delete(@tqent_id[tid]);
+}
+
+END
+{
+        clear(@birth);
+        clear(@start);
+
+        clear(@tq_name);
+        clear(@tqent_id);
+}
+EOF
+
+KVER=$(uname -r)
+bpftrace \
+        --include "/usr/src/zfs-$KVER/zfs_config.h" \
+        -I "/usr/src/zfs-$KVER/include" \
```

It's kind of annoying to have this bash wrapper that doesn't really do that much.  I think we should consider making a standalone script that would invoke bpftrace with the right arguments to use with ZFS (possible names `zbpf` or `bpftrace_zfs`).  Then your script could use `#!/usr/bin/zbpf`

---

I agree; but I'm not sold on the addition of some new script that's a wrapper around `bpftrace`, but adds the `-I` flag.. I'd like to keep what I have here for now, until we have a solid plan for how to address this moving forward; then we can retrofit this script to use whatever mechanism we decided long term.

For example, I wonder if it'd be better to augment `bpftrace` itself to look for headers in the ZFS directory when the directory exists, but I don't know how feasible that would be without having more time to investigate that solution.

I'd like to avoid rushing to whatever solution is easiest to implement (now), without first trying to think about the most correct solution, and pursuing that first.

---

I haven't tested this, but I've opened https://github.com/delphix/bpftrace/pull/4 .. Let me know what you think, and if I should spend time to test and verify such an approach would work.

---

This probably isn't the correct place to have this discussion, but it made me remember and realize that we only keep the "latest" set of headers installed on the system at any given time. So, for example, if we install a new set of headers but don't reboot to start running the modules that correspond to those headers, I think this approach may not work (I don't know if it'll work by "accident", silently return the wrong information, or explicitly fail).

---

The upgrade-but-not-rebooted case is interesting, thanks for bringing that up.

If the kernel was also updated, then I think everything would "just work" since `uname -r` / `utsname.release` and thus the path to the zfs includes would be different for the newer version, and we'd keep using the older include files.

But if the kernel version was exactly the same, then the new ZFS headers would be copied over the old ones, right?  In which case we'd be using the newer headers with the older (still running) kernel module.

Would it be possible to include the Delphix release in `uname -r` somehow?  Or should we change the path to include the ZFS module "version" somehow?

---

I had opened a bug about this a while back, here's a link to it: https://jira.delphix.com/browse/DLPX-65703